### PR TITLE
Update to camera ready version

### PR DESCRIPTION
**Pull request by koenhelwegen**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -56,13 +56,22 @@ bnno train binarynet \
...
+bnno train xnornet --dataset imagenet2012 --hparams-set bop
+bnno train birealnet --dataset imagenet2012 --hparams-set bop
 ```
+
+This should give the following results:
+
+| Network | Bop - top-1 accuracy |
+|----|----|
+|Binary Alexnet | 41.1% |
+|XNOR-architecture | 45.9%*  |
```
Comment (lgeiger) [inline README.md]: Either add the "*No scaling factors used" or remove the `*` 
```suggestion
|XNOR-architecture | 45.9% |
```

File: bnn_optimization/data/cifar10.py [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
+from bnn_optimization.data.problem_definitions import ImageClassification
+import tensorflow as tf
 
 
 @registry.register_preprocess("cifar10")
-def default(image):
-    return tf.cast(image, tf.float32) / (255.0 / 2.0) - 1.0
+class default(ImageClassification):
+    def inputs(self, data):
+        return super().inputs(data) / (255.0 / 2.0) - 1.0
```
Comment (lgeiger) [inline bnn_optimization/data/cifar10.py]: Do we use this preprocessing anywhere? If not we can remove it

File: bnn_optimization/data/imagenet.py [hunk 3]
```diff
@@ -1,235 +1,38 @@
...
 
-    Returns:
-      A preprocessed image.
-    """
-    if is_training:
-        # For training, we want to randomize some of the distortions.
-        resize_side = tf.random.uniform(
-            [], minval=resize_side_min, maxval=resize_side_max + 1, dtype=tf.int32
+@registry.register_preprocess("imagenet2012")
+class vgg(ImageNetClassification):
```
Comment (lgeiger) [inline bnn_optimization/data/imagenet.py]: We don't use this anywhere so remove.

File: bnn_optimization/data/imagenet.py [hunk 4]
```diff
@@ -1,235 +1,38 @@
...
-        resize_side = resize_side_min
-        crop_fn = _central_crop
 
-    num_channels = image.get_shape().as_list()[-1]
-    image = _aspect_preserving_resize(image, resize_side)
-    image = crop_fn(image, output_height, output_width)
 
-    image.set_shape([output_height, output_width, num_channels])
+@registry.register_preprocess("imagenet2012")
+class efficientnet_augment(ImageNetClassification):
```
Comment (lgeiger) [inline bnn_optimization/data/imagenet.py]: We don't use this --> 🔥

File: bnn_optimization/data/preprocessing/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from bnn_optimization.data.preprocessing import efficientnet, vgg
+
+__all__ = ["efficientnet", "vgg"]
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/__init__.py]: We don't use `vgg` so we should remove it. Also should we rename the efficientnet preprocessing to something else since it is basically renset preprocessing and it might be confusing since the efficientnet paper uses autoaugment by default and we don't

File: bnn_optimization/data/preprocessing/autoaugment.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
+"""AutoAugment policies for enhanced image preprocessing.
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/autoaugment.py]: We should remove this.

File: bnn_optimization/data/preprocessing/efficientnet.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
+"""ImageNet preprocessing.
+
+Code from https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/preprocessing.py
+"""
+import tensorflow as tf
+from bnn_optimization.data.preprocessing import autoaugment
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/efficientnet.py]: Since we don't use autoaugment, we should remove it

File: bnn_optimization/data/preprocessing/efficientnet.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        image = tf.image.random_flip_left_right(image)
+    else:
+        image = _decode_and_center_crop(image_bytes, image_size)
+
+    image = tf.reshape(image, [image_size, image_size, 3])
+
+    if is_training and autoaugment_name:
+        image = tf.clip_by_value(image, 0.0, 255.0)
+        image = tf.cast(image, dtype=tf.uint8)
+        image = autoaugment.distort_image_with_autoaugment(image, autoaugment_name)
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/efficientnet.py]: See above

File: bnn_optimization/data/preprocessing/efficientnet.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+    image = tf.cast(image, dtype=tf.bfloat16 if use_bfloat16 else tf.float32)
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/efficientnet.py]: Should we cleanup all `bfloat16` stuff, since we are only running on `float32`?

File: bnn_optimization/data/preprocessing/vgg.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
+"""Provides utilities to preprocess images.
```
Comment (lgeiger) [inline bnn_optimization/data/preprocessing/vgg.py]: We don't use it so remove

File: bnn_optimization/models/birealnet.py [hunk 11]
```diff
@@ -66,42 +68,108 @@ def residual_block(x, double_filters=False, filters=None):
             out = residual_block(out)
 
     # layer 18
-    out = tf.keras.layers.GlobalAvgPool2D()(out)
-    out = tf.keras.layers.Dense(dataset.num_classes, activation="softmax")(out)
+    if include_top:
```
Comment (lgeiger) [inline bnn_optimization/models/birealnet.py]: This is only used in larq zoo, so I think we should remove the check

File: bnn_optimization/models/birealnet.py [hunk 12]
```diff
@@ -66,42 +68,108 @@ def residual_block(x, double_filters=False, filters=None):
...
     kernel_initializer = "glorot_normal"
 
     @property
     def optimizer(self):
-        decay_step = 100 * 1281167 // self.batch_size
-        lr = tf.keras.optimizers.schedules.PolynomialDecay(
-            2.5e-3, decay_step, end_learning_rate=2.5e-6, power=1.0
-        )
+        if self.decay_schedule == "linear_cosine":
+            lr = tf.keras.experimental.LinearCosineDecay(self.learning_rate, 750684)
```
Comment (lgeiger) [inline bnn_optimization/models/birealnet.py]: Not used in the experiments

File: bnn_optimization/models/birealnet.py [hunk 13]
```diff
@@ -66,42 +68,108 @@ def residual_block(x, double_filters=False, filters=None):
...
+    @property
+    def optimizer(self):
+        decay_step = self.epochs * 1281167 // self.batch_size
+        lr = tf.keras.optimizers.schedules.PolynomialDecay(
+            self.lr_start, decay_step, end_learning_rate=self.lr_end, power=1.0
+        )
+        gamma = tf.keras.optimizers.schedules.PolynomialDecay(
+            self.gamma_start, decay_step, end_learning_rate=self.gamma_end, power=1.0
+        )
+        threshold = tf.keras.optimizers.schedules.PolynomialDecay(
```
Comment (lgeiger) [inline bnn_optimization/models/birealnet.py]: If I remember correctly, we don't decay the threshold so please remove as well.

File: bnn_optimization/models/birealnet.py [hunk 14]
```diff
@@ -66,42 +68,108 @@ def residual_block(x, double_filters=False, filters=None):
...
+        )
+        return lq.optimizers.Bop(
+            tf.keras.optimizers.Adam(0.0 * lr),
+            threshold=1e5 + threshold,
+            gamma=0.0 * gamma,
+        )
+
+
+@registry.register_hparams(birealnet)
+class eval(bop):
```
Comment (lgeiger) [inline bnn_optimization/models/birealnet.py]: Is this used anywhere?

File: bnn_optimization/models/xnor.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    @property
+    def optimizer(self):
+        decay_step = self.epochs * 1281167 // self.batch_size
+        lr = tf.keras.optimizers.schedules.PolynomialDecay(
+            self.lr_start, decay_step, end_learning_rate=self.lr_end, power=1.0
+        )
+        gamma = tf.keras.optimizers.schedules.PolynomialDecay(
+            self.gamma_start, decay_step, end_learning_rate=self.gamma_end, power=1.0
+        )
+        threshold = tf.keras.optimizers.schedules.PolynomialDecay(
```
Comment (lgeiger) [inline bnn_optimization/models/xnor.py]: See above

File: bnn_optimization/optimizers.py [hunk 16]
```diff
@@ -1,37 +1,93 @@
...
-                "fp_optimizer is not an object of tf.keras.optimizers.Optimizer"
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
             )
 
-        self._fp_optimizer = fp_optimizer
+        self.fp_optimizer = fp_optimizer
         self._set_hyper("threshold", threshold)
         self._set_hyper("gamma", gamma)
 
+        logger = logging.Logger("rethink_logger")
```
Comment (lgeiger) [inline bnn_optimization/optimizers.py]: I think the logger should be registered outside the `__init__`

File: bnn_optimization/optimizers.py [hunk 17]
```diff
@@ -1,37 +1,93 @@
-import larq as lq
 import tensorflow as tf
+import larq as lq
+
+from larq import utils
```
Comment (lgeiger) [inline bnn_optimization/optimizers.py]:
```suggestion
```

File: bnn_optimization/optimizers.py [hunk 18]
```diff
@@ -1,37 +1,93 @@
...
+from copy import deepcopy
 
+import logging
 
+
+@utils.register_keras_custom_object
```
Comment (lgeiger) [inline bnn_optimization/optimizers.py]:
```suggestion
@lq.utils.register_keras_custom_object
```

File: requirements.txt [hunk 19]
```diff
@@ -0,0 +1,6 @@
+click==7.0
+tensorflow-datasets==1.2.0
+polyaxon-client==0.5.6
```
Comment (lgeiger) [inline requirements.txt]:
```suggestion
```

File: requirements.txt [hunk 20]
```diff
@@ -0,0 +1,6 @@
...
+larq==0.7.4
+zookeeper==0.5.5
+tensorflow-addons==0.6.0
```
Comment (lgeiger) [inline requirements.txt]: If we remove the autoaugment code, we can remove this too. Also this would break with TF 1
```suggestion
```

File: setup.py [hunk 21]
```diff
@@ -18,15 +24,10 @@ def readme():
...
-        "tensorflow-datasets==1.0.2",
-        "larq==0.2.0",
-        "zookeeper==0.1.1",
-    ],
+    install_requires=install_requires(),
     extras_require={
-        "tensorflow": ["tensorflow==1.14.0rc0"],
-        "tensorflow_gpu": ["tensorflow-gpu==1.14.0rc0"],
+        "tensorflow": ["tensorflow==1.14.0"],
+        "tensorflow_gpu": ["tensorflow-gpu==1.14.0"],
```
Comment (lgeiger) [inline setup.py]:
```suggestion
        "tensorflow_gpu": ["tensorflow-gpu==2.0.0"],
```

File: setup.py [hunk 22]
```diff
@@ -18,15 +24,10 @@ def readme():
...
-        "click==7.0",
-        "tensorflow-datasets==1.0.2",
-        "larq==0.2.0",
-        "zookeeper==0.1.1",
-    ],
+    install_requires=install_requires(),
     extras_require={
-        "tensorflow": ["tensorflow==1.14.0rc0"],
-        "tensorflow_gpu": ["tensorflow-gpu==1.14.0rc0"],
+        "tensorflow": ["tensorflow==1.14.0"],
```
Comment (lgeiger) [inline setup.py]: We ran the latest experiments with TF 2, correct?

```suggestion
        "tensorflow": ["tensorflow==2.0.0"],
```

File: bnn_optimization/models/binary_alexnet.py [hunk 23]
```diff
@@ -0,0 +1,93 @@
+import tensorflow as tf
+import larq as lq
+from zookeeper import registry, HParams
+
+
+@registry.register_model
+def binary_alexnet(hparams, input_shape, num_classes):
```
[the remark this answers is not in the document]
Comment (koenhelwegen) [reply bnn_optimization/models/binary_alexnet.py]: The details of the networks are quite different so I would definitely prefer two separate models. We could share the building blocks but tbh I think this is more trouble than it's worth (also, when reading this code as a researcher I prefer to have the complete model definition in a single place).

File: bnn_optimization/models/binary_alexnet.py [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+    return tf.keras.Model(inputs=img_input, outputs=out)
+
+
+@registry.register_hparams(binary_alexnet)
+class bop(HParams):
+    epochs = 150
+    batch_size = 1024
+    input_quantizer = "ste_sign"
+    kernel_quantizer = tf.keras.layers.Activation("linear")
```
[the remark this answers is not in the document]
Comment (koenhelwegen) [reply bnn_optimization/models/binary_alexnet.py]: Bop targets the weights with quantizers, so unfortunately we cannot remove this.

File: bnn_optimization/models/birealnet.py [hunk 25]
```diff
@@ -67,41 +69,41 @@ def residual_block(x, double_filters=False, filters=None):
...
     filters = 64
-    batch_size = 256
+    learning_rate = None
+    decay_schedule = "linear"
+    epochs = 300
+    batch_size = 1024
     input_quantizer = "approx_sign"
-    kernel_quantizer = "magnitude_aware_sign"
-    kernel_constraint = "weight_clip"
+    kernel_quantizer = tf.keras.layers.Activation("linear")
```
[the remark this answers is not in the document]
Comment (koenhelwegen) [reply bnn_optimization/models/birealnet.py]: see above

File: bnn_optimization/models/birealnet.py [hunk 26]
```diff
@@ -67,41 +69,41 @@ def residual_block(x, double_filters=False, filters=None):
...
+    threshold = 1e-8
 
+    gamma_start = 1e-4
+    gamma_end = 1e-6
 
-@registry.register_hparams(birealnet)
-class bop(default):
-    kernel_quantizer = None
-    kernel_constraint = None
+    lr_start = 2.5e-3
```
[the remark this answers is not in the document]
Comment (koenhelwegen) [reply bnn_optimization/models/birealnet.py]: I prefer this but happy to change it

File: README.md [hunk 27]
```diff
@@ -56,13 +56,22 @@ bnno train binarynet \
...
 ```
-bnno train birealnet --dataset imagenet2012 --hparams-set bop --epochs 100
+bnno train alexnet --dataset imagenet2012 --hparams-set bop
+bnno train xnornet --dataset imagenet2012 --hparams-set bop
+bnno train birealnet --dataset imagenet2012 --hparams-set bop
 ```
+
+This should give the following results:
+
+| Network | Bop - top-1 accuracy |
```
[the remark this answers is not in the document]
Comment (koenhelwegen) [reply README.md]: No would keep the info here strictly about our reported results. Context is provided in the paper.
